include/ensmallen_bits/cmaes/cmaes_impl.hpp

```diff
@@ -313,6 +325,90 @@ typename MatType::elem_type CMAES<SelectionPolicyType>::Optimize(
   return overallObjective;
 }
 
+// Transforms the candidate into the given bounds.
+template<typename SelectionPolicyType>
+template<typename BaseMatType>
+void CMAES<SelectionPolicyType>::BoundaryTransform(BaseMatType& matrix)
+{
+  typedef typename BaseMatType::elem_type ElemType;
+  const double diff = (upperBound - lowerBound) / 2.0;
+  const double al = std::min(diff, (1 + std::abs(lowerBound)) / 20.0);
+  const double au = std::min(diff, (1 + std::abs(upperBound)) / 20.0);
+  const double xlow = lowerBound - 2 * al - diff;
+  const double xup = upperBound + 2 * au + diff;
+  const double r = 2 * (upperBound - lowerBound + al + au);
+
+  for (size_t col = 0; col < matrix.n_cols; col++)
+  {
+    for (size_t row = 0; row < matrix.n_rows; row++)
+    {
+      ElemType y = matrix(row, col);
+      // Boundary transformation shift into feasible pre-image.
+      if (y < xlow)
+      {
+        y += (ElemType)(r * (1 + (xlow - y) / r));
+      }
+      else if (y > xup)
+      {
+        y -= (ElemType)(r * (1 + (y - xup) / r));
+      }
+      else if (y < lowerBound - al)
+      {
+        y += (ElemType)(2 * (lowerBound - al - y));
+      }
+      else if (y > upperBound + au)
+      {
+        y -= (ElemType)(2 * (y - upperBound - au));
+      }
+      // Boundary transformation.
+      if (y < lowerBound + al)
+      {
+        y = (ElemType)(lowerBound + (y - (lowerBound - al)) *
+            (y - (lowerBound - al)) / 4.0 / al);
+      }
+      else if (y > upperBound - au)
+      {
+        y = (ElemType)(upperBound - (y - (upperBound + au)) *
+            (y - (upperBound + au)) / 4.0 / au);
+      }
+
+      matrix(row, col) = y;
+    }
+  }
+}
+
+// Computes the inverse of the transformation.
+template<typename SelectionPolicyType>
+template<typename BaseMatType>
+void CMAES<SelectionPolicyType>::BoundaryTransformInverse(BaseMatType& matrix)
+{
+  typedef typename BaseMatType::elem_type ElemType;
+  const double diff = (upperBound - lowerBound) / 2.0;
+  const double al = std::min(diff, (1 + std::abs(lowerBound)) / 20.0);
+  const double au = std::min(diff, (1 + std::abs(upperBound)) / 20.0);
+
+  for (size_t col = 0; col < matrix.n_cols; col++)
+  {
+    for (size_t row = 0; row < matrix.n_rows; row++)
+    {
+      ElemType y = matrix(row, col);
+      
+      if (y < lowerBound + al)
+      {
+        y =  (ElemType)(lowerBound - al) + 2 *
+              std::pow(std::abs(al * (y - lowerBound)),0.5);
```

Comment:
Any reason to avoid `sqrt`?

Reply:
Sorry, @zoq for this mistake I have rectified it. Have a look at it now.

Reply:
Thanks!